RA_Integration/IRA_Dialog.cpp

```diff
@@ -0,0 +1,320 @@
+#include <Windows.h>
+#include <WindowsX.h>
+#include <CommCtrl.h>
+
+#include "IRA_Dialog.h"
+#include "RA_Core.h"
+
+long IRA_Dialog::storage{ 0 };
+
+// Note to self: Never call DefWindowProc or DefDlgProc for dialogs
+
+IRA_Dialog::IRA_Dialog(int resId, bool isModal) :
+	ResourceId{ resId },
+	is_modal{ isModal }
+{
+	storage = reinterpret_cast<long>(this);
+}
+
+IRA_Dialog::~IRA_Dialog()
+{
+	storage = 0;
+}
+
+INT_PTR IRA_Dialog::DoModal()
+{
+	return DialogBox(g_hThisDLLInst, MAKEINTRESOURCE(ResourceId), g_RAMainWnd, MsgQueue);
+}
+
+HWND IRA_Dialog::Create()
+{
+	return CreateDialog(g_hThisDLLInst, MAKEINTRESOURCE(ResourceId), g_RAMainWnd, MsgQueue);
+}
+
+INT_PTR IRA_Dialog::MsgQueue(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+	IRA_Dialog* ira = reinterpret_cast<IRA_Dialog*>(storage);
```

Comment:
isn't `ira` just going to be `this`? why obscure it and store it as a static variable in the class instance?

Reply:
A static function needs an instance to a class. Since we need a pure virtual function for Callbacks between child and parent windows this is necessary, since you cannot make an instance of an abstract class. "IRA_Dialog" technically does not exist, it's child dialogs and controls do.

Reply:
If this was for a WindowProc we wouldn't need this and just set it to nullptr and intilizate during WM_NCCREATE. Since this is a Dialog different rules apply and the window is already supposed to exist but g_RAMainWnd is a standalone handle.

For the dialog, the pointer cannot be null, so we give it something so it won't be null. It's not a big deal since it only exists in virtual space anyway.

Reply:
again, didn't realize the method was static.